record_function: add torchbind alternative API

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #72303
* #72302
* __->__ #72301

First step in resolving #35026.

This adds `PythonRecordFunction` which is a `torch::CustomClassHolder`
for `at::RecordFunction` to keep the ATen code free of torch includes.
And adds new unused internal API functions
`_record_function_enter_new` which return the torchbind object.

Once the FC period is expired, `torch.profiler.record_function` will
be updated to use this new internal API. Then once BC period is
expired, the cpp_custom_type_hack-based API can be removed.

Differential Revision: [D34586311](https://our.internmc.facebook.com/intern/diff/D34586311)

peterbell10 committed:
record_function: add torchbind alternative API

First step in resolving #35026.

This adds `PythonRecordFunction` which is a `torch::CustomClassHolder`
for `at::RecordFunction` to keep the ATen code free of torch includes.
And adds new unused internal API functions
`_record_function_enter_new` which return the torchbind object.

Once the FC period is expired, `torch.profiler.record_function` will
be updated to use this new internal API. Then once BC period is
expired, the cpp_custom_type_hack-based API can be removed.

[ghstack-poisoned]

peterbell10 committed:
Update on "record_function: add torchbind alternative API"


First step in resolving #35026.

This adds `PythonRecordFunction` which is a `torch::CustomClassHolder`
for `at::RecordFunction` to keep the ATen code free of torch includes.
And adds new unused internal API functions
`_record_function_enter_new` which return the torchbind object.

Once the FC period is expired, `torch.profiler.record_function` will
be updated to use this new internal API. Then once BC period is
expired, the cpp_custom_type_hack-based API can be removed.

[ghstack-poisoned]

peterbell10 committed:
Update on "record_function: add torchbind alternative API"


First step in resolving #35026.

This adds `PythonRecordFunction` which is a `torch::CustomClassHolder`
for `at::RecordFunction` to keep the ATen code free of torch includes.
And adds new unused internal API functions
`_record_function_enter_new` which return the torchbind object.

Once the FC period is expired, `torch.profiler.record_function` will
be updated to use this new internal API. Then once BC period is
expired, the cpp_custom_type_hack-based API can be removed.

[ghstack-poisoned]

peterbell10 committed:
Update on "record_function: add torchbind alternative API"


First step in resolving #35026.

This adds `PythonRecordFunction` which is a `torch::CustomClassHolder`
for `at::RecordFunction` to keep the ATen code free of torch includes.
And adds new unused internal API functions
`_record_function_enter_new` which return the torchbind object.

Once the FC period is expired, `torch.profiler.record_function` will
be updated to use this new internal API. Then once BC period is
expired, the cpp_custom_type_hack-based API can be removed.

[ghstack-poisoned]

peterbell10 committed:
Update on "record_function: add torchbind alternative API"


First step in resolving #35026.

This adds `PythonRecordFunction` which is a `torch::CustomClassHolder`
for `at::RecordFunction` to keep the ATen code free of torch includes.
And adds new unused internal API functions
`_record_function_enter_new` which return the torchbind object.

Once the FC period is expired, `torch.profiler.record_function` will
be updated to use this new internal API. Then once BC period is
expired, the cpp_custom_type_hack-based API can be removed.

[ghstack-poisoned]

peterbell10 committed:
Update on "record_function: add torchbind alternative API"


First step in resolving #35026.

This adds `PythonRecordFunction` which is a `torch::CustomClassHolder`
for `at::RecordFunction` to keep the ATen code free of torch includes.
And adds new unused internal API functions
`_record_function_enter_new` which return the torchbind object.

Once the FC period is expired, `torch.profiler.record_function` will
be updated to use this new internal API. Then once BC period is
expired, the cpp_custom_type_hack-based API can be removed.

[ghstack-poisoned]

peterbell10 committed:
Rebase on master on "record_function: add torchbind alternative API"


First step in resolving #35026.

This adds `PythonRecordFunction` which is a `torch::CustomClassHolder`
for `at::RecordFunction` to keep the ATen code free of torch includes.
And adds new unused internal API functions
`_record_function_enter_new` which return the torchbind object.

Once the FC period is expired, `torch.profiler.record_function` will
be updated to use this new internal API. Then once BC period is
expired, the cpp_custom_type_hack-based API can be removed.

[ghstack-poisoned]

albanD commented:
cc @robieta can you review this one?

robieta reviewed (approved):
LGTM. (Sorry for the delay.)

robieta commented:
@robieta has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34586311).

robieta commented on torch/csrc/autograd/record_function_ops.cpp:
```diff
@@ -67,36 +90,71 @@ c10::intrusive_ptr<c10::ivalue::Future> _call_end_callbacks_on_fut(
+TORCH_LIBRARY_FRAGMENT(profiler, m) {
+  m.class_<PythonRecordFunction>("_RecordFunction");
+
+  m.def("_record_function_enter", &record_function_enter_legacy);
```
This needs to be `_record_function_enter(str name, str? args=None) -> Tensor`, as there are internal calls which are only providing `name` and failing because the second arg no longer has a default value.

peterbell10 committed:
Include full schema for _record_function_enter on "record_function: add torchbind alternative API"


First step in resolving #35026.

This adds `PythonRecordFunction` which is a `torch::CustomClassHolder`
for `at::RecordFunction` to keep the ATen code free of torch includes.
And adds new unused internal API functions
`_record_function_enter_new` which return the torchbind object.

Once the FC period is expired, `torch.profiler.record_function` will
be updated to use this new internal API. Then once BC period is
expired, the cpp_custom_type_hack-based API can be removed.

Differential Revision: [D34586311](https://our.internmc.facebook.com/intern/diff/D34586311)

[ghstack-poisoned]

robieta commented:
I reran the tests but they're still failing. Can you rebase?

peterbell10 committed:
Rebase on master on "record_function: add torchbind alternative API"


First step in resolving #35026.

This adds `PythonRecordFunction` which is a `torch::CustomClassHolder`
for `at::RecordFunction` to keep the ATen code free of torch includes.
And adds new unused internal API functions
`_record_function_enter_new` which return the torchbind object.

Once the FC period is expired, `torch.profiler.record_function` will
be updated to use this new internal API. Then once BC period is
expired, the cpp_custom_type_hack-based API can be removed.

Differential Revision: [D34586311](https://our.internmc.facebook.com/intern/diff/D34586311)

[ghstack-poisoned]

peterbell10 committed:
Fix clang-format violation on "record_function: add torchbind alternative API"


First step in resolving #35026.

This adds `PythonRecordFunction` which is a `torch::CustomClassHolder`
for `at::RecordFunction` to keep the ATen code free of torch includes.
And adds new unused internal API functions
`_record_function_enter_new` which return the torchbind object.

Once the FC period is expired, `torch.profiler.record_function` will
be updated to use this new internal API. Then once BC period is
expired, the cpp_custom_type_hack-based API can be removed.

Differential Revision: [D34586311](https://our.internmc.facebook.com/intern/diff/D34586311)

[ghstack-poisoned]

peterbell10 commented:
Rebased and CI is now passing.

robieta commented:
@robieta has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34586311).
